File: Data Structures & Algorithms/combination-target-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        dfs(res, nums, {}, 0, target, 0);
        return res;
    }

    void dfs(vector<vector<int>>& res, vector<int>& nums, vector<int> subset, int index, int target, int currSum) {

        if (currSum == target) {
            res.push_back(subset);
            return;
        }

        if (index >= nums.size() || currSum > target) {
            return;
        }

        dfs(res, nums, subset, index + 1, target, currSum);
        subset.push_back(nums[index]);
        dfs(res, nums, subset, index, target, currSum + nums[index]);
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        i = 0
        while i < len(nums):
            l = i + 1
            r = len(nums) - 1

            while l < r:
                total = nums[i] + nums[l] + nums[r]

                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
            
            i += 1
            while i < len(nums) and nums[i] == nums[i - 1]:
                i += 1
        
        return res
